gr-fec/lib/fec_mtrx_impl.cc

```diff
@@ -396,42 +403,56 @@ gsl_matrix* fec_mtrx_impl::calc_inverse_mod2(const gsl_matrix* original_matrix)
     gsl_matrix_scale(matrix_inverse, determinant);
 
     // Take mod 2 of each element in the matrix.
-    for (row_index = 0; row_index < n; row_index++) {
-        for (col_index = 0; col_index < n; col_index++) {
+    for (unsigned int row_index = 0; row_index < n; row_index++) {
+        for (unsigned int col_index = 0; col_index < n; col_index++) {
 
-            float value = gsl_matrix_get(matrix_inverse, row_index, col_index);
+            const auto value = gsl_matrix_get(matrix_inverse, row_index, col_index);
 
             // take care of mod 2
-            int value_cast_as_int = static_cast<int>(value);
-            int temp_value = abs(fmod(value_cast_as_int, 2));
+            // FIXME: we cast to int, which will be problematic when something is
+            // x.9999998; then we do an fmod on that int; then we do an abs on the float
+            // that comes out of that, which we then truncate to int.
+            //
+            // Not many ways this could be less efficient, and more numerically
+            // questionable.
+            const int value_cast_as_int = static_cast<int>(value);
+            const int temp_value = abs(fmod(value_cast_as_int, 2));
```

Comment:
Can we go and fix this in this PR? I agree that it is not a long-term solution, but this is a bug. It's probably not too difficult to find matrices for which this gives the wrong inverse because of the flooring.

I would do something like `uint8_t temp_value = static_cast<int>(std::round(value)) % 2`, but I never remember how `% 2` (or `& 1`) work for negative numbers. This might need an `abs`.

Reply:
yes, we can; I was avoiding putting it in the same commit, but will fix that this weekend.

Reply:
For me the problem is not remembering how 2's complement work, but the fact that C and C++ have nasty surprises whenever signed integers are involved (specially if you start doing bitwise operations with them).

In this case, it seems that `& 1` does the desired thing (just a bitwise AND), while `% 2` does something else: https://godbolt.org/z/PvW3EnGj1